kube-runtime/src/manager.rs

```diff
@@ -0,0 +1,99 @@
+where
+    K: Clone + Meta + DeserializeOwned + 'static,
+{
+    /// Create a ControllerBuilder on a type `K`
+    ///
+    /// Configure `ListParams` and `Api` so you only get reconcile events
+    /// for the correct Api scope (cluster/all/namespaced), or ListParams subset
+    ///
+    /// A writer is exposed for convenience so you can peak into the main reflector's state.
+    pub fn new(owned_api: Api<K>, lp: ListParams, writer: Writer<K>) -> Self {
+        let reader = writer.as_reader();
+        let mut selector = vec![];
+        let self_watcher: Pin<Box<dyn Stream<Item = Result<ObjectRef<K>, watcher::Error>>>> = Box::pin(
+            trigger_self(try_flatten_applied(reflector(writer, watcher(owned_api, lp)))),
+        );
+        selector.push(self_watcher);
+        Self { selector, reader }
+    }
+
+    /// Indicate child objets K owns and be notified when they change
+    ///
+    /// This type `CHILD` must have OwnerReferences set to point back to `K`.
+    /// You can customize the parameters used by the underlying watcher if
+    /// only a subset of `CHILD` entries are required.
+    /// The `api` must have the correct scope (cluster/all namespaces, or namespaced)
+    pub fn owns<CHILD: Clone + Meta + DeserializeOwned + 'static>(
+        mut self,
+        api: Api<CHILD>,
+        lp: ListParams,
+    ) -> Self {
+        let child_watcher = trigger_owners(try_flatten_touched(watcher(api, lp)));
+        self.selector.push(Box::pin(child_watcher));
+        self
+    }
+
+    /// Indicate an object to watch with a custom mapper
+    ///
+    /// This mapper should return something like Option<ObjectRef<K>>
+    pub fn watches<
+        OTHER: Clone + Meta + DeserializeOwned + 'static,
+        I: 'static + IntoIterator<Item = ObjectRef<K>>,
+    >(
+        mut self,
+        api: Api<OTHER>,
+        lp: ListParams,
+        mapper: impl Fn(OTHER) -> I + 'static,
+    ) -> Self {
+        let other_watcher = trigger_with(try_flatten_touched(watcher(api, lp)), mapper);
+        self.selector.push(Box::pin(other_watcher));
+        self
+    }
+
+    /// Consume the ControllerBuilder and start the controller stream
+    ///
+    /// This creates a stream from all builder calls and starts a controller with
+    /// a specified `reconciler` and `error_policy` callbacks. Each of these will be called
+    /// with your configurable `Context`.
+    pub fn run<ReconcilerFut, T>(
+        self,
+        reconciler: impl FnMut(K, Context<T>) -> ReconcilerFut,
+        error_policy: impl FnMut(&ReconcilerFut::Error, Context<T>) -> ReconcilerAction,
+        context: Context<T>,
+    ) -> impl Stream<Item = Result<(ObjectRef<K>, ReconcilerAction), Error<ReconcilerFut::Error, watcher::Error>>>
+    where
+        K: Clone + Meta + 'static,
+        ReconcilerFut: TryFuture<Ok = ReconcilerAction>,
+        ReconcilerFut::Error: std::error::Error + 'static,
+    {
+        let input_stream = stream::select_all(self.selector);
```

Comment:
`futures::stream::SelectAll` is already a mutable collection, so we can build it immediately rather than going via the `Vec` first.

Reply:
i would have to figure out the type of the SelectAll though, and that was the hardest part that caused me to go the vector route.

Reply:
Should be `SelectAll<Pin<Box<dyn Stream<Item = _>>>` from what I can tell, just like the current `Vec`.

Reply:
Thanks that works. Still not able to omit the extra type hint (because it's too opaque for it), but at least we now avoid the vector.